components/MapFilters.vue

```diff
@@ -0,0 +1,262 @@
+          </span>
+        </div>
+      </div>
+      <br>
+    </div>
+  </div>
+</template>
+
+<script>
+ export default {
+   props: {
+     startingMarkers: {
+       type: Array,
+       required: true
+     },
+     filterLayers: {
+       type: Array,
+       required: true
+     },
+     isLoading: {
+       type: Boolean,
+       required: true
+     }
+   },
+   data () {
+     return {
+       filters: false,
+       noFiltersOn: 0,
+       markers: [],
+       originalMarkers: [],
+       currentTileName : 'Modern Map',
+       tiles: [{name: 'Modern Map', url: 'https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png', active: true},
+               {name: 'Historic Map', url: 'https://api.maptiler.com/tiles/uk-osgb1919/{z}/{x}/{y}.jpg?key=cKVGc9eOyhb8VH5AxCtw', active : false}],
+       currentLayer: 0, // Index in filterLayers corresponding to current layer.
+     }
+   },
+   watch: {
+     startingMarkers: {
+       handler(newValue, oldValue) {
+         this.originalMarkers = JSON.parse(JSON.stringify(newValue));
+         this.markers = JSON.parse(JSON.stringify(newValue));
+       }
+     }
+   },
+   methods: {
+     getMarkerType : function (marker) {
+       // Returns the marker type. If it detects that
+       // one witch has a different type to the others,
+       // returns mixed straight away. Otherwise returns
+       // markerType, which will be common to all witches.
+
+       let markerType = '';
+       let witches = marker.witches;
+
+       for (let i = 0, len = witches.length; i < len; i++) {
+         let witch = witches[i];
+         let witchType = witch[this.currentFilterProperty];
+
+         if (i === 0) {
+           markerType = witchType;
+         } else if (witchType !== markerType) {
+           return 'mixed';
+         }
+       }
+
+       return markerType;
+     },
+     getMarkerIcon: function (marker) {
+       // Sets the incon of a marker depending on its
+       // marker type, which it gets from calling
+       // getMarkerType.
+
+       let markerType = this.getMarkerType(marker);
+
+       if (markerType === 'mixed') {
+         return '/images/witch-single-purple.png';
+       } else {
+         return this.filterLayers[this.currentLayer].filters[markerType].iconUrl;
+       }
+     },
+     updateMarkerState: function (marker) {
+       // Updates the state of a marker after it being
+       // filtered. Specifically, its marker Icon and
+       // if it needs to be on or off depending on whether
+       // all the witches have been filtered out.
+
+       if (marker.witches.length > 0) {
+         marker.markerIcon = this.getMarkerIcon(marker);
+         marker.onOff = true;
+       } else {
+         marker.onOff = false;
+       }
+
+       return marker;
+     },
+     removeWitchesFromArray: function (witches, filterType) {
+       // Given an array of witches, it filters out the
+       // witches that meet the property filterType.
+
+       let fwitches = [];
```

Comment:
It is probably better to name fwitches something like filteredWitches, just so it is a bit more obvious what the variable is holding. We try to avoid acronyms.